lib/httpoison/base.ex

```diff
@@ -432,6 +433,29 @@ defmodule HTTPoison.Base do
      end
   end
 
+  defp do_request(method, request_url, request_headers, {:stream, enumerable}, hn_options) do
+    with {:ok, ref} <- :hackney.request(method, request_url, request_headers, :stream, hn_options) do
+
+      failures = Stream.transform(enumerable, :ok, fn
+        _, :error -> {:halt, :error}
+        bin, :ok  -> {[], :hackney.send_body(ref, bin)}
+        _, error  -> {[error], :error}
+      end) |> Enum.into([])
```

Comment:
Using `Enum.into` here to build a list of failures. On the happy path, it only emits empty list to `Stream.transform` after sending the body, so we're not materializing the list before sending. On the failure side it emits the error, and then changes the state to `:error` so the error can get emitted and then the stream halts on the next go round. 

This is why I'm not using `Stream.run` here - we want the failure back in the case that something goes wrong.